src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -196,6 +198,224 @@ class segment_meta_column_frame {
+/// possible.
+template<class value_t, class delta_t>
+class segment_meta_column;
+
+/// Column implementatoin
+///
+/// Contains a list of frames. The frames are not overlapping with each other.
+/// The iterator is used to scan both all frames seamlessly.
+/// This specialization is for xor-delta algorithm. It doesn't allow skipping
+/// frames so all search operations require full scan. Random access by index
+/// can skip frames since indexes are monotonic.
+template<class value_t>
+class segment_meta_column<value_t, details::delta_xor> {
+    using frame_t = segment_meta_column_frame<value_t, details::delta_xor>;
+    using decoder_t = deltafor_decoder<value_t, details::delta_xor>;
+    static constexpr size_t max_frame_size = 0x1000;
+
+public:
+    using const_iterator
+      = segment_meta_column_const_iterator<value_t, details::delta_xor>;
+
+    using delta_alg = details::delta_xor;
+
+    explicit segment_meta_column(delta_alg d)
+      : _delta_alg(d) {}
+
+    void append(value_t value) {
+        if (_frames.empty() || _frames.back().size() == max_frame_size) {
+            _frames.emplace_back(_delta_alg);
+        }
+        _frames.back().append(value);
+    }
+
+    const_iterator at(size_t index) const {
+        auto ix_outer = index / max_frame_size;
+        auto ix_inner = index % max_frame_size;
+        auto it = _frames.begin();
+        std::advance(it, ix_outer);
+        return const_iterator(it, _frames.end(), ix_inner);
+    }
+
+    size_t size() const {
+        if (_frames.empty()) {
+            return 0;
+        }
+        return max_frame_size * (_frames.size() - 1) + _frames.back().size();
+    }
+
+    const_iterator begin() const { return const_iterator(_frames); }
+
+    const_iterator end() const { return const_iterator(); }
+
+    const_iterator find(value_t value) const {
+        return pred_search<std::equal_to<value_t>>(value);
+    }
+
+    const_iterator upper_bound(value_t value) const {
+        return pred_search<std::greater<value_t>>(value);
+    }
+
+    const_iterator lower_bound(value_t value) const {
+        return pred_search<std::greater_equal<value_t>>(value);
+    }
+
+    std::optional<value_t> last_value() const {
+        if (size() == 0) {
+            return std::nullopt;
+        }
+        return _frames.back().last_value();
+    }
+
+    bool contains(value_t value) const { return find(value) != end(); }
+
+    /// Prefix truncate the column
+    ///
+    /// \param new_start is value from which the column should start
+    /// \note the method can only be used for monotonic sequences
+    void prefix_truncate(value_t new_start) {
+        auto st = _frames.begin();
+        for (auto it = _frames.begin(); it != _frames.end(); it++) {
+            if (it->last_value() < new_start) {
+                st = it;
+            } else {
+                it->prefix_truncate(new_start);
+                break;
+            }
+        }
+        _frames.erase(_frames.begin(), st);
```

Comment:
I'm not sure what kind of container frames is, but in std containers erase is usually non-inclusive on the right side of the interval. assuming that is true for frames, do you want `std::next(st)`? I ask because it seems like when `st = it` above, the condition indicates that you'd want `it` to be removed too since its last value is less than new start?